Character controller

Chris-Nicholls opened the pull request:
Started to expose some character controller functionality.

Not all the functionality is exposed yet, just enough to create and move a capsule controller.

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+    }
+}
+
+fn from_extended(vec: PxExtendedVec3) -> Vec3 {
```
Are these functions used outside the Character Controller? If so, I think they belong in either `transform.rs` or `math/mod.rs`.

Chris-Nicholls replied:
No, the extended types seem to only be used for character controllers.

tgolsson replied:
Great, then this makes sense for now!

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+    filters: PxControllerFilters,
+}
+
+impl Controller {
```
Can we use the `[physx_type]` macro pattern here? It's not straight forward to combine it with data ownership, unfortunately. Maybe it'd make sense to separate this into two parts, a wrapper around the PhysX-sys API (the trait) and the actual controller? It feels like the combination of these two data pieces might belong more at the application level. (I.e., in my specific character implementation for my game). 

So something like:

```rust
#[px_type]
impl Controller {
    // all API functions for PxController 
}

struct Controllable {
    controller: Controller,
    filters: ControllerFilters,
}
```

Chris-Nicholls replied:
Ok, I wasn't quite sure what that macro did. I think I get it now. 

Regarding Controller/Controllable, would it make sense to use Deref and DerefMut? 
```impl Deref for Controllable {
    type Target = Controller;
    fn deref(&self) -> &Controller {
        &self.controller
    }
}

impl DerefMut for Controllable {
    fn deref_mut(&mut self) -> &mut Controller {
        &mut self.controller
    }
}
```

tgolsson replied:
No, but I'm not sure if such a `Controllable` belongs in physx-rs at all since it's not a PxSomething (right?). We only use Deref/DerefMut to emulate direct inherited function calls since there is no mechanic to do that nicely in Rust. In this case, there is nothing called a PxControllable that _is-a_ PxController as far as I know. But the `physx_type` macro can do it for the subtypes of `PxController`: 

```rust
#[physx_type(inherit = "Controller")]
impl CapsuleController {
    // PxCapsuleControllerAPI
};
```

This'll implement appropriate `GetRaw`, `Deref` and `DerefMut` for `PxType<PxCapsuleController>` and alias it as CapsuleController.

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    pub fn get_raw_mut(&mut self) -> *mut PxController {
```
Can we implement this using the `GetRaw` trait instead?

Chris-Nicholls replied:
👍 Didn't see that

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+
+    pub fn set_position(&mut self, position: Vec3) {
+        unsafe {
+            PxController_setPosition_mut(self.get_raw_mut(), &(to_extended(&position)));
```
Is the extra set of parenthesis required here? (I'm guessing clippy would complain if it wasn't, but just checking since it looks odd).

Chris-Nicholls replied:
Uh, no. I'll remove them.

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    pub fn move_by(&mut self, disp: &Vec3, min_distance: f32, elapsed_time: f32) {
```
I think we can take the `disp` parameter directly by value here, since the first thing we do with it is deref it.

Chris-Nicholls replied:
👍

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+        unsafe { from_extended(*PxController_getPosition(self.get_raw())) }
+    }
+
+    pub fn get_actor(&self) -> *mut PxRigidDynamic {
```
Can this return a Rust level RigidDynamic instead?

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,93 @@
+    }
+}
+
+impl Drop for Controller {
```
I'd prefer if we did not have automatic release at this level, because you can easily wreck your application:

```rust
let player_position = Controller::new(self.controller_manager.get_controller(0)).get_position(); // boom now, or maybe later
```

Chris-Nicholls replied:
Good point.

tgolsson commented on physx/src/scene.rs:
```diff
@@ -88,6 +91,39 @@ impl Scene {
+        step_offset: f32,
+    ) -> Controller {
+        unsafe {
+            if self.controller_manager.is_none() {
```
I think it makes more sense to put this as part of the scene builder API, not as a runtime opt-in thing. That also lets the user decide whether to use locking.

tgolsson replied:
Also, does this require any extensions to be loaded?

Chris-Nicholls replied:
Sure, that makes sense. Would still you leave it as optional? 
And no, this doesn't require any extensions.

tgolsson replied:
I think having it as optional makes sense, as there's lots of use-cases where you probably don't use it. Should be fine as a `use_controller_manager(enable: bool, locking: bool)` on the builder.

tgolsson commented on physx/src/scene.rs:
```diff
@@ -88,6 +91,39 @@ impl Scene {
+                ));
+            }
+
+            let c = PxCapsuleControllerDesc_new_alloc();
```
If you're up for it, I'd prefer if this was an actual rust struct as well, so we stick closer to the PhysX api. That also makes it easier if we end up exposing things like Materials as Rust-level types. Then we can also get rid of the panic, and let the caller decide what to do if the struct wasn't constructed correctly.

If you know, I'm also curious about the ownership here - this returns a pointer, but it's not clear to me if the controller or controller manager takes ownership of it later.

Chris-Nicholls replied:
Looking at it again, I don't think either take ownership - they just read the data. So it should be ok to delete it straight after using it.

Chris-Nicholls replied:
Ok, I had a go at converting into proper rust struct. Let me know what you think!

Chris-Nicholls commented:
Hey @tgolsson I had a go at addressing your comments. Thanks for your time in reviewing this - I'm quite new to rust, so I appreciate you taking a look.

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,103 @@
+
+    pub fn create_controller(&self, desc: &mut ControllerDesc) -> Controller {
+        let controller =
+            unsafe { PxControllerManager_createController_mut(self.ptr, desc.get_raw_mut()) };
```
The desc here should be just `get_raw()` here, it's a `*const PxControllerDesc`.

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,103 @@
+            if PxCapsuleControllerDesc_isValid(c) {
+                Ok(CapsuleControllerDesc { ptr: c })
+            } else {
+                Err(format!(
```
Returning strings as errors is suboptimal, since it's hard to match on them on the consumer side. The most common approach is to return an enum that implements std::error::Error, like this:

```rust
pub mod controller {
    pub enum Error {
        InvalidDesc { height: f32, radius: f32, offset: f32 }
    }
    impl std::error::Error for Error {
        ...
    }

    type Result<T> = std::result::Result<T, Error>;
}

fn maybe_succeeds() -> controller::Result<Something> { ... }
```

There's a crate called `thiserror` that's commonly used to make this a bit nicer: https://docs.rs/thiserror/1.0.11/thiserror/. It should be quite straightforward to convert your String errors use this library instead, but I'll happily provide more info or help out if you want.

Also, we should free the pointer in the failure branch as well, so we don't leak it.

Chris-Nicholls replied:
Ok, added the error enum

tgolsson commented on physx/src/scene.rs:
```diff
@@ -88,6 +91,31 @@ impl Scene {
+        height: f32,
+        radius: f32,
+        step_offset: f32,
+    ) -> Result<Controller, String> {
```
If this function takes the description instead, we don't have to return an error here. :)

Chris-Nicholls replied:
Unfortunately, it still needs one for the case when no controller manager has been added.

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,115 @@
+#[derive(Error, Debug)]
+pub enum ControllerError {
+    #[error("Controller description is invalid")]
+    InvalidDescription(),
```
This can be `InvalidDescription { height: f32,  ... }` etc, and then you can refer to those in the error just like in your original string. It makes it a bit more descriptive. :)

tgolsson commented on physx/src/controller.rs:
```diff
@@ -0,0 +1,115 @@
+    }
+
+    pub fn get_actor(&self) -> RigidDynamic {
+        unsafe { RigidDynamic::new(PxController_getActor(self.ptr)) }
```
use `from_ptr` here instead. They have slightly different semantics that we should probably make more apparent. `new` will allocate a Box of UserData, while `from_ptr` does not. :)

tgolsson commented on physx/src/scene.rs:
```diff
@@ -503,6 +538,11 @@ impl SceneBuilder {
         self
     }
 
+    pub fn use_controller_manager(&mut self, use_controller_manager: bool) -> &mut Self {
```
If we could also pass in the controller manager locking flag here that'd be lovely!

tgolsson commented on physx/src/scene.rs:
```diff
@@ -520,7 +560,7 @@ impl SceneBuilder {
 
     /// Build a new Scene from the provided parameters
-    pub(super) fn build(&self, physics: &mut Physics) -> PxSceneDesc {
+    pub(super) fn build(&self, physics: &mut Physics) -> Scene {
```
This is much cleaner, thanks! :)

tgolsson commented on physx/src/scene.rs:
```diff
@@ -88,6 +96,23 @@ impl Scene {
 
+    pub fn add_capsule_controller(
+        &mut self,
+        desc: &mut CapsuleControllerDesc,
```
Should be fine with a non-mutable borrow here, I think? :)

tgolsson commented:
@Chris-Nicholls! We're hoping to get a new release out on Thursday with this change in. Thanks! I'm hoping to review your other two PRs so they can also go out in the same version.

hrydgard commented:
Discovered too late that this gets rid of the only way to get a raw *PxScene, which is sometimes needed when you do stuff that the wrapper doesn't yet cover. So will have to make a followup patch release that adds get_raw to Scene.
